unison-cli/src/Unison/LSP.hs

```diff
@@ -0,0 +1,158 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE TypeOperators #-}
+
+module Unison.LSP where
+
+import Colog.Core (LogAction (LogAction))
+import qualified Colog.Core as Colog
+import Control.Monad.Reader
+import Data.Aeson hiding (Options, defaultOptions)
+import qualified Ki
+import qualified Language.LSP.Logging as LSP
+import Language.LSP.Server
+import Language.LSP.Types
+import Language.LSP.Types.SMethodMap
+import qualified Language.LSP.Types.SMethodMap as SMM
+import Language.LSP.VFS
+import qualified Network.Simple.TCP as TCP
+import Network.Socket
+import System.Environment (lookupEnv)
+import Unison.Codebase
+import Unison.Codebase.Branch (Branch)
+import qualified Unison.Codebase.Path as Path
+import Unison.Codebase.Runtime (Runtime)
+import qualified Unison.LSP.FileAnalysis as Analysis
+import Unison.LSP.Hover (hoverHandler)
+import Unison.LSP.NotificationHandlers as Notifications
+import Unison.LSP.Orphans ()
+import Unison.LSP.Types
+import Unison.LSP.UCMWorker (ucmWorker)
+import qualified Unison.LSP.VFS as VFS
+import Unison.Parser.Ann
+import Unison.Prelude
+import qualified Unison.PrettyPrintEnvDecl as PPED
+import Unison.Symbol
+import UnliftIO
+
+getLspPort :: IO String
+getLspPort = fromMaybe "5757" <$> lookupEnv "UNISON_LSP_PORT"
+
+-- | Spawn an LSP server on the configured port.
+spawnLsp :: Codebase IO Symbol Ann -> Runtime Symbol -> STM (Branch IO, Path.Absolute) -> IO ()
+spawnLsp codebase runtime ucmState = withSocketsDo do
+  lspPort <- getLspPort
+  TCP.serve (TCP.Host "127.0.0.1") lspPort $ \(sock, _sockaddr) -> do
+    sockHandle <- socketToHandle sock ReadWriteMode
+    Ki.scoped \scope -> do
```

Comment:
It's a bit of an anti-pattern to pass a `scope` down into a function; could this scope be created closer to where it's used, or not really?

Reply:
I don't think it can, the goal of this scope is for every background thread and request which is related to a given LSP connection should be tied to the scope for that connection, so that if the client ever hangs up, all the background work terminates as well. Since this is where the thread will "hang" waiting for a hang-up, I think this is the right place to put it, but correct me if you can think of a better way 😄